### ISSUE TITLE
Allow to compose layers of Irmin storages

### ISSUE THREAD

**Issue by samoht** [mirage/irmin#885]
There are various design options, but the idea is to have something like overlayFS but for Irmin layers. All the layers are read-only but the last one which is read-write.

We could expose a `freeze` operation which change the last layer to read-only and could perform an GC to only keep live objects.

This could interact nicely with naturally layered backends, like `irmin-pack` with its index where it's important to keep the writeable part of the system as small as possible.

**Comment by icristescu** [comment]
I have worked a bit on this and I think that there are some design choices to make, of which I'm not sure what is the best approach. 

One is the naming scheme for the layers. In `irmin-pack` a name identifies a store and if the "freezed" layer is a separate store it needs a different name. This can also be true for `irmin` in general, using the `root` of `Conf`.
I've investigated two approaches for this : 
- Let the user handle the naming scheme by asking for a name when [moving to another layer](https://github.com/mirage/irmin/pull/882/commits/8a5abbec9a9ec7e25c48463b68365166ddc331c3#diff-73b958dbfc5f4f8687a16fb58ca66baaR100). 
- Hide the naming scheme from the user. Only [freeze](https://github.com/mirage/irmin/pull/882/commits/2eef67a651cdc399191a230fa7a7e6463bb4af94#diff-1be34172140401295ad7258b1add6782R3216) is exposed to the user. Internally only a single layer is used, named `root^"_layered"`. `freeze` adds the contents of the current store to the `root^"_layered"` but has to close the current store and open a new one if it wants a clean store. And here the naming issue reappears: do we generate new names and let the user delete the old files from the disk? or do we remove the old files and reuse the same name? Note that the first suggestion also complicates opening a store after a close.    

Another is the tag store. I think that the assumption here is that the tag store is [small enough to keep in memory](https://github.com/mirage/irmin/blob/master/src/irmin-pack/irmin_pack.ml#L173). Does it make sense to freeze this type of store?
